Automattic/wp-calypso
# SSR: Whitelist (and validate) redirect_to for the /log-in page SSR

Before this commit, accessing a page like https://wordpress.com/log-in?redirect_to=https://wordpress.com/themes would have not been Server-Side-Rendered (only signup_flow & client_id were considered valid URL params for this). This is an issue because we are interested in generating a valid canonical URL for the pages that are indexed in search engines (Google, Bing, Yandex), and some of these search engines require a server-side-rendered HTML.

One way to easily enable SSR (which is currently strongly linked to the cache), is to add `redirect_to` to the `context.cacheQueryKeys` array. As we don't want to fill the cache with absolutely everything that may come through redirect_to, an additional `context.cacheQueryCallableValidators` was added which represents a hashmap (key to callable) for validating different URL (parameters. In case of redirect_to, only the URLs starting with https://wordpress.com/theme and https://wordpress.com/go would be cached (these are also the URLs that are currently reported as invalid by our SEO audit).

Semi-related note: an interesting aspect that I noticed is that `client_id` actually requires to have a `redirect_to` defined (otherwise the flow throws an exception). But, once redirect_to AND client_id are defined (which is always true when client_id is present), then the pages are not SSRed (because redirect_to was not on the list of whitelisted parameters). After this commit, nothing will change in this behavior, because only redirect_to that start with https://wordpress.com would be SSRed.

Testing:
1) Checkout the branxh, start npm with SSR debugging: `DEBUG='calypso:server-render' npm start`
2) Access http://calypso.localhost:3000/log-in - you should see the page as SSRed, with valid canonical and title (Log-In). Refresh the page and expect the same.
3) Access http://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Fgo%2Fweb-design%2Fleave-it-to-the-pros-outsource-your-website-security%2F and look at the Calypso debug logs. You should expect to see some SSR cache hit infos, but nothing that would say that URL is not cacheable.
4) Modify something in the URL above such that it doesn't start with wordpress.com (i.e. replace wordpress.com with bluehost). You should see in the logs a message saying that 'URL not cacheable because of failing the contraints for key redirect_to'.
5) Add one parameter to the URL in step 3. You should see that 'SSR: URL is not cacheable because it contains additional keys than the whitelist'
6) Repeat 3, 4 and 5 with two more URLs (one of them starting with themes also). Example: 1) https://wordpress.com/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Ftags%2F & 2) https://wordpress.com/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Fthemes%2Ffilter%2Ffeature%3Avideo%2F
7) Look at the code and think of ways of breaking it.
